src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java

```diff
@@ -0,0 +1,100 @@
+import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.PositionUtils;
+import com.redhat.devtools.intellij.lsp4mp4ij.psi.core.utils.PsiTypeUtils;
+import com.redhat.devtools.intellij.quarkus.psi.internal.builditems.QuarkusBuildItemErrorCode;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4mp.commons.DocumentFormat;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class QuarkusBuildItemDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
+
+    private static final String BUILD_ITEM= "io.quarkus.builder.item.BuildItem";
+
+    @Override
+    public boolean isAdaptedForDiagnostics(JavaDiagnosticsContext context) {
+        // Collection of diagnostics for Quarkus Build Items is done only if
+        // io.quarkus.builder.item.BuildItem is on the classpath
+        Module javaProject = context.getJavaProject();
+        return PsiTypeUtils.findType(javaProject, BUILD_ITEM) != null;
+    }
+
+    @Override
+    public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context) {
+        PsiFile typeRoot = context.getTypeRoot();
+        PsiElement[] elements = typeRoot.getChildren();
+        List<Diagnostic> diagnostics = new ArrayList<>();
+        collectDiagnostics(elements, diagnostics, context);
+        return diagnostics;
+    }
+
+    private static void collectDiagnostics(PsiElement[] elements, List<Diagnostic> diagnostics,
+                                           JavaDiagnosticsContext context) {
+        for (PsiElement element : elements) {
+            if (element instanceof PsiClass) {
+                PsiClass psiClass = (PsiClass) element;
+                if (isBuildItem(psiClass)) {
+                    validateBuildItem(psiClass, diagnostics, context);
+                }
+            }
+        }
+    }
+
+    private static boolean isBuildItem(PsiClass type) {
+        return InheritanceUtil.isInheritor(type, BUILD_ITEM);
+    }
+
+    private static void validateBuildItem(PsiClass psiClass, List<Diagnostic> diagnostics, JavaDiagnosticsContext context) {
+        if (psiClass.hasModifierProperty(PsiModifier.FINAL)
+        || psiClass.hasModifierProperty(PsiModifier.ABSTRACT)
+        ) {
+            return;
+        }
+        Range range = PositionUtils.toClassDeclarationRange(psiClass, context.getUtils());
+        Diagnostic d = context.createDiagnostic(context.getUri(),
+                createDiagnosticMessage(psiClass, context.getDocumentFormat()),
+                range, "quarkus",
+                QuarkusBuildItemErrorCode.InvalidModifierBuildItem,
+                DiagnosticSeverity.Error
+        );
+        diagnostics.add(d);
+    }
+
+    private static String createDiagnosticMessage(PsiClass classType, DocumentFormat documentFormat) {
+        StringBuilder message = new StringBuilder("BuildItem class ");
```

Comment:
Use a message format (by declaring the key message as static)

Reply:
@fbricon I agree that using à prooérties file is cleaner but we need to think about translation in JDT. Using a static string key for message will avoid translating this code

Reply:
done